### PR TITLE
[lang] Support matrix initialization with a list of vectors

### PR DESCRIPTION
**Pull request by KLozes**
Related issue = #810 

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by KLozes** [comment]
I think the rows/cols is a good idea. Should we make it mandatory? And should it also be used for list of list style initialization?

**Comment by yuanming-hu** [comment]
> I think the rows/cols is a good idea. Should we make it mandatory? And should it also be used for list of list style initialization?

Yeah I think we can make it mandatory for making matrices from vectors. For list of list, let's keep the old style, since changing that might break some code.

Btw, I had to do a force-push just now to fix a PR merging issue on the master branch - could you update your local `master` branch, and `rebase` (or `cherrypick`) commit 919bfdbe422a475aad14b9eff054f834d7312bbb based on it, and then (force-)update your `vec_to_matrix` branch? Sorry about the trouble...

**Comment by archibate** [comment]
Sorry about the confusing commits! To remove them:
```bash
git checkout master
git reset --hard origin/master
git pull --set-upstream origin master
git checkout vector_to_matrix
git rebase -i master
```
Then you'll see:
```
pick xxx [glsl] xxx # delete this line!
pick xxx [glsl] yyy # delete this line!
pick xxx [lang] zzz
pick xxx [lang] www
```
Please delete those lines that are not related to this PR, then save and quit (`:wq`).

**Comment by yuanming-hu** [comment]
> Sorry about the confusing commits!

That was my fault - when merging #785 I accidentally chose `Create a merge commit` instead of `Squash and merge`. Then I had to use a force push to fix that, leading to the divergence here. 

To prevent this from happening again, I have disabled `Create a merge commit` and `Rebase and merge` in this repo...

**Comment by KLozes** [comment]
oh no. lots and lots of test failing with, not just test_ad_atomic 

```
test_ad_atomic.py F                                                                                                 [100%]

======================================================== FAILURES =========================================================
_____________________________________________________ test_ad_reduce ______________________________________________________

>   ???

/home/klozes/Documents/software/taichi/tests/python/test_ad_atomic.py:27: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../python/taichi/lang/kernel.py:484: in wrapped
    primal(*args, **kwargs)
../../python/taichi/lang/kernel.py:414: in __call__
    self.materialize(key=key, args=args, arg_features=arg_features)
../../python/taichi/lang/kernel.py:232: in materialize
    src = remove_indent(inspect.getsource(self.func))
/usr/lib/python3.6/inspect.py:973: in getsource
    lines, lnum = getsourcelines(object)
/usr/lib/python3.6/inspect.py:955: in getsourcelines
    lines, lnum = findsource(object)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

object = <function test_ad_reduce.<locals>.func at 0x7f50f83bdb70>

    def findsource(object):
        """Return the entire source file and starting line number for an object.
    
        The argument may be a module, class, method, function, traceback, frame,
        or code object.  The source code is returned as a list of all the lines
        in the file and the line number indexes a line in that list.  An OSError
        is raised if the source code cannot be retrieved."""
    
        file = getsourcefile(object)
        if file:
            # Invalidate cache if needed.
            linecache.checkcache(file)
        else:
            file = getfile(object)
            # Allow filenames in form of "<something>" to pass through.
            # `doctest` monkeypatches `linecache` module to enable
            # inspection, so let `linecache.getlines` to be called.
            if not (file.startswith('<') and file.endswith('>')):
                raise OSError('source code not available')
    
        module = getmodule(object, file)
        if module:
            lines = linecache.getlines(file, module.__dict__)
        else:
            lines = linecache.getlines(file)
        if not lines:
>           raise OSError('could not get source code')
E           OSError: could not get source code

/usr/lib/python3.6/inspect.py:786: OSError
```

File: docs/linalg.rst [hunk 1]
```diff
@@ -17,4 +17,145 @@ Matrices
 - ``R, S = ti.polar_decompose(A, ti.f32)``
 - ``U, sigma, V = ti.svd(A, ti.f32)`` (Note that ``sigma`` is a ``3x3`` diagonal matrix)
 
-TODO: doc here better like Vector.
+TODO: doc here better like Vector. WIP
+
+A matrix in Taichi can have two forms:
+
+  - as a temporary local variable. An ``n by m`` matrix consists of ``n * m`` scalar values.
+  - as a an element of a global tensor. In this case, the tensor is an ND array of ``n by m`` matrices.
```
Comment (yuanming-hu) [inline docs/linalg.rst:25]: Would `N-dimensional` be easier to understand than `ND`?

File: docs/linalg.rst [hunk 2]
```diff
@@ -17,4 +17,145 @@ Matrices
...
+        a = ti.Matrix([2, 3, 4])
+
+.. note::
+
+    this is equivalent to ti.Vector([x, y, ...])
+
+
+.. function:: ti.Matrix([[x, y, ...], [z, w, ...], ...])
+
+    :parameter x: (scalar) the first component of the first column
```
Comment (yuanming-hu) [inline docs/linalg.rst:72]: Actually, by default, `[x, y]` would be the elements of the first **row** :-)

This makes `[[x, y], [z, w]]` consistent with `list` of lists, `numpy` ndarrays and math notations $M_{i,j}$: the first index is for the row, the second the column.

File: docs/linalg.rst [hunk 3]
```diff
@@ -17,4 +17,145 @@ Matrices
...
+    :parameter y: (scalar) the second component of the first column
+    :parameter z: (scalar) the first component of the second column
+    :parameter w: (scalar) the second component of the second column
+
+    For example, this creates a 2x2 matrix with components (2, 3) in the first column and (4, 5) in the second column:
```
Comment (yuanming-hu) [inline docs/linalg.rst:77]: Same here

File: docs/vector.rst [hunk 4]
```diff
@@ -5,9 +5,8 @@ Vectors
 
 A vector in Taichi can have two forms:
 
-  - as a temporary local variable, an ``n``-D vector consists of ``n`` scalar values.
-  - as a global tensor, where each tensor element is a vector. In this case, an ``n``-D vector consists of ``n`` global tensors of scalars.
-    The tensors of scalars, when treated together, can be considered to be **a global tensor of vectors**.
+  - as a temporary local variable. An ``n`` component vector consists of ``n`` scalar values.
+  - as an element of a global tensor. In this case, the tensor is an ND array of ``n`` component vectors
```
Comment (yuanming-hu) [inline docs/vector.rst:9]: Again, ND array might be ambiguous for people who don't use `numpy` :-)

**Comment by KLozes** [comment]
oh wait I see why its failing. for some reason the path to the test.py is wrong. but not for all tests

File: python/taichi/lang/matrix.py [hunk 5]
```diff
@@ -19,16 +19,36 @@ class Matrix:
...
                  shape=None,
                  layout=None,
                  needs_grad=False,
-                 keep_raw=False):
+                 keep_raw=False,
+                 rows=None,
+                 cols=None):
         self.grad = None
-        if isinstance(n, list):
+        if rows is not None or cols is not None:
```
Comment (yuanming-hu) [inline python/taichi/lang/matrix.py:33]: Let's be slightly more defensive here: Raise a `ValueError` if
 - `rows` and `cols` present simultaneously;
 - elements of `rows` or `cols` do not have the same shape;
 - elements of `rows` or `cols` do not have the `m == 1`.

**Comment by KLozes** [comment]
bug fixed. all test seem to be running fine. I just had to delete the __pycache__ in the test directory. I change the local repo name last week and I guess the cached files remembered the old name.

File: docs/linalg.rst (hunk 1 above)
Comment (KLozes) [reply docs/linalg.rst:25]: I like N-dimensional, lets go with that.

**Pull request merged**